pldmtool/pldmtool.cpp

```
@@ -90,23 +115,221 @@ class MctpRawOp : public CommandInterface
         app->add_option("-d,--data", rawData, "raw MCTP data")
             ->required()
             ->expected(-3);
+        app->add_option("-i,--file-in ", inFileName, "Read in the file to be sent");
+        app->add_option("-o,--file-out", outFileName, "Write out the received file");
+        app->add_flag("-c,--checksum", checksum, "Append/Validate checksum");
         app->add_flag("-p,--prealloc-tag", mctpPreAllocTag,
                       "use pre-allocated MCTP tag for this request");
     }
-    std::pair<int, std::vector<uint8_t>> createRequestMsg() override
-
-    {
-        return {PLDM_SUCCESS, rawData};
-    }
-
-    void parseResponseMsg(pldm_msg* /* responsePtr */,
-                          size_t /* payloadLength */) override
-    {}
 
+    const std::string getInFileName() const {return inFileName;}
+    const std::string getOutFileName() const {return outFileName;}
+    std::pair<int, std::vector<uint8_t>> createRequestMsg() override;
+    void parseResponseMsg(pldm_msg *, size_t) override;
   private:
     std::vector<uint8_t> rawData;
+    std::string outFileName{};
+    std::string inFileName{};
+    bool checksum{false};
+
+    void appendInfile();
 };
 
+/*
+ * @brief Append an input firmware file to the rawData buffer for transmission.
+ *
+ * This function:
+ *   1. Opens the input file specified by getInFileName().
+ *   2. Reads the entire file into memory.
+ *   3. Appends the file-size field (big-endian) to rawData.
+ *   4. Appends the file contents to rawData.
+ *   5. Clears the IC (Integrity Check) bit in the header, and if checksum
+ *      is enabled, sets it and appends a CRC32 checksum of the file.
+ *
+ * All errors (file open/read failures) are printed and cause an early return.
+ */
+
+void MctpRawOp::appendInfile()
```

NIT:
Use std::filesystem::file_size() instead of tellg()

Why: More reliable and clear for getting file size; avoids stream state issues and works even if file is opened later.



Enable ifstream.exceptions()

Why: Converts silent failures into exceptions, making error handling uniform and reducing manual checks.



Stream file in chunks (instead of reading whole file into memory)

Why: Prevents huge allocations for large files, improves scalability and memory efficiency.



Reserve capacity in rawData

Why: Avoids repeated reallocations when appending size, file data, and checksum.



Use std::byteswap + std::endian for endian conversion

Why: Idiomatic C++23 approach; eliminates manual bit-twiddling and improves readability.



Use std::span for CRC computation

Why: Safer and clearer than raw pointers; expresses intent without pointer arithmetic.